C#: Store nullability on the side

calumgrant opened the pull request:
This is a subset of https://github.com/Semmle/ql/pull/1606 that deals only with type nullability, and some very small changes to the dbscheme.

- Refactor nullability to include an on-the-side data structure to encode the nullability. This models the way that C#8 makes nullability backwards compatible, which is that the nullability is actually encoded using attributes rather than in the type itself. The reason for this change is to not break the database when dealing with projects that are compiled without nullability, or in situations where the assembly is compiled with nullability information and is referenced by another assembly that does not use nullability.
- Update the database stats.
- Replace `varchar` to `string` in the dbscheme.

hvitved commented on csharp/ql/src/semmle/code/csharp/AnnotatedType.qll:
```diff
@@ -147,32 +243,47 @@ private int getBitMask(int bit) {
-      type = c.getTypeArgument(i) and
-      annotations = Annotations::TAnnotationFlags(getTypeArgumentFlags(c, i))
+    exists(AnnotatedConstructedType c, int i |
+      type = c.getType().(ConstructedType).getTypeArgument(i) and
```
Add
```
override ConstructedType getType()
```
to the class `AnnotatedConstructedType` and get rid of the infix cast.

calumgrant replied:
Good suggestion but it introduces a negative recursion that doesn't really seem worthwhile addressing.

hvitved commented on csharp/ql/src/semmle/code/csharp/AnnotatedType.qll:
```diff
@@ -147,32 +243,47 @@ private int getBitMask(int bit) {
+    annotations.getNoFlagsNullability() instanceof Annotations::NoNullability
+    or
+    exists(AnnotatedArrayType at |
+      type = at.getType().(ArrayType).getElementType() and
```
Add
```
override ArrayType getType()
```
to the class `Annotated ArrayType` and get rid of the infix cast.

calumgrant replied:
This also introduced a negative recursion.

hvitved commented on csharp/ql/src/semmle/code/csharp/AnnotatedType.qll:
```diff
@@ -233,14 +344,70 @@ class AnnotatedType extends TAnnotatedType {
   /** Holds if this annotated type applies to element `e`. */
   predicate appliesTo(Element e) { Annotations::elementTypeAnnotations(e, type, annotations) }
 
+  /** Holds if this annotated type is the `ith` type argument of constructed generic 'g'. */
```
`` `ith` `` -> `` `i`th ``.

hvitved commented on csharp/ql/src/semmle/code/csharp/AnnotatedType.qll:
```diff
@@ -10,129 +10,225 @@ import csharp
     }
 
+    /** Gets the nullability with no additional flags. */
+    Nullability getNoFlagsNullability() { this = TAnnotationsNoFlags(result) }
```
This seems like an internal implementation detail that should not be exposed.

calumgrant commented on csharp/ql/src/semmle/code/csharp/AnnotatedType.qll:
```diff
@@ -10,129 +10,225 @@ import csharp
+   * A structured type annotation representing type nullability.
+   * For example, `IDictionary<string!,string?>?` has nullability `<!,?>?`.
+   */
+  abstract class Nullability extends TypeAnnotation, TNullability {
```
It doesn't need to be `abstract`. `getBit()` was abstract but it's better just to avoid `abstract` and use `none()` instead.

calumgrant commented on csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.cs:
```diff
@@ -345,11 +345,10 @@ protected void PopulateGenerics(TextWriter trapFile)
                         child++;
                     }
+
+                    trapFile.type_nullability(this, NullabilityEntity.Create(Context, new Nullability(symbol)));
```
Good idea, thanks!

calumgrant commented on csharp/extractor/Semmle.Extraction.CSharp/Entities/Symbol.cs:
```diff
@@ -28,11 +28,13 @@ protected void PopulateAttributes()
-            var ta = annotation.GetTypeAnnotation();
-            if (ta != Kinds.TypeAnnotation.None)
-                trapFile.type_annotation(this, ta);
+            var n = NullabilityEntity.Create(Context, Nullability.Create(type));
```
This is the place where the fully oblivious type annotation gets emitted to the trap file. So we'd want to extract this in all circumstances. (It cached and only extracted once per file.)

calumgrant commented on csharp/ql/src/semmle/code/csharp/Type.qll:
```diff
@@ -821,34 +821,31 @@ class ArrayType extends DotNet::ArrayType, RefType, @array_type {
   }
 
-  private string getRankString(int i) {
+  string getRankString(int i) {
```
It's used by `AnnotatedArrayType` so I think exposing it is reasonable.

hvitved replied:
Right, but it is an internal implementation detail; I wouldn't expect users of the library to be able to call this predicate.

calumgrant replied:
Alternatives are: 1) Commenting `/** INTERNAL: Do not use ... */` 2) pulling the code out into a `library` predicate somewhere 3) Duplicating the predicate in `AnnotatedArrayType` 4) Leaving as is. I'm leaning towards 1) but take your pick.

hvitved replied:
Let's go for 1) then.

calumgrant commented on csharp/ql/test/library-tests/tostringwithtypes/toStringWithTypes.expected:
```diff
@@ -1,7 +1,7 @@
 | Delegate<> | Delegate<T>(int, T) |
-| Delegate<String> | Delegate<string> |
```
`toStringWithTypes()` now gives the same string as `toString()` so the result disappeared from the query.

calumgrant commented on csharp/ql/src/semmle/code/csharp/AnnotatedType.qll:
```diff
@@ -147,32 +243,47 @@ private int getBitMask(int bit) {
+  if specific_type_parameter_nullability(constraints, getTypeRef(type), _)
+  then specific_type_parameter_nullability(constraints, getTypeRef(type), result.getNullability())
+  else (
+    specific_type_parameter_constraints(constraints, type) and
```
I think so.

hvitved commented on csharp/ql/src/semmlecode.csharp.dbscheme:
```diff
@@ -480,8 +480,8 @@ case @nullability.kind of
-#keyset[nullability, index]
-nullability_member(int nullability: @nullability ref, int index: int ref, int child: @nullability ref)
+#keyset[parent, index]
+nullability_parent(int nullability: @nullability ref, int index: int ref, int parent: @nullability ref)
```
Add `unique` in front of `int nullability: @nullability ref`.

calumgrant replied:
They're not unique! The same nullability can have many parents. e.g. the nullability of `!` has parents `<!,!>!` and `<!,!,!>!`. So I could have pushed back on this suggestion, as calling something `nullability_parent` suggests a single parent, when really you only think about navigating from a nullabilty to one of its children. Still I've renamed it to `nullability_parent` so that might as well stay.

hvitved replied:
You are right! Sorry about the confusion.

calumgrant commented:
# Overview
- time: Thu Nov 14 2019 10:19:01 GMT+0000 (UTC)
- host: profiling-csharp-calum.northeurope.cloudapp.azure.com
- dir: /tmp/dist-compare__APRO
- shas: cbdb97070c2@39a45ceef cbdb97070c2@18d613832
- projects: 12
- options: --suite /home/calum/code/target/docker/intree/profiling-dist-csharp/queries/customer/default/csharp-lgtm

## Report generation messages:
- Warning: /tmp/dist-compare__APRO/sarif/monodevelop_81d2012344f8a7e603c7eaf0066cad880af6266c_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/monodevelop_81d2012344f8a7e603c7eaf0066cad880af6266c_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/corefx_510ab9e6433ff0df29e0ae080140f4f2c1b69abd_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/corefx_510ab9e6433ff0df29e0ae080140f4f2c1b69abd_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/EntityFrameworkCore_a2f17c7f90b63f57b56fc2a336e115b7b1e960d3_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/EntityFrameworkCore_a2f17c7f90b63f57b56fc2a336e115b7b1e960d3_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/runuo_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/runuo_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/coreclr_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/coreclr_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/PowerShell_ba94c6cfeb93d66bd30748ffa181d2e6d541351e_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/PowerShell_ba94c6cfeb93d66bd30748ffa181d2e6d541351e_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/pythonnet_51a186858096577aef6a860656523d3194eee9a8_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/pythonnet_51a186858096577aef6a860656523d3194eee9a8_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/roslyn_6a52142bef1db0a0936c483d13f426acae335555_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/roslyn_6a52142bef1db0a0936c483d13f426acae335555_revision-cbdb97070c2@18d613832.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/nhibernate-core_7997e22b1d999fecacd91dc8664f5132001cebfb_revision-cbdb97070c2@39a45ceef.sarif does not exist.
- Warning: /tmp/dist-compare__APRO/sarif/nhibernate-core_7997e22b1d999fecacd91dc8664f5132001cebfb_revision-cbdb97070c2@18d613832.sarif does not exist.
# Timing
```
Project                                               cbdb97070c2@39a45ceef  cbdb97070c2@18d613832
OneOf                                                 1206                   1171                   0.970978
OrchardCore_999eb06bfd9a263881b98bdd135ee54a33002c55  1337                   1401                   1.04787
ql                                                    479                    506                    1.05637

Snapshots with query failures:
  none
Snapshots with build failures:
  monodevelop_81d2012344f8a7e603c7eaf0066cad880af6266c/revision-cbdb97070c2@39a45ceef
  monodevelop_81d2012344f8a7e603c7eaf0066cad880af6266c/revision-cbdb97070c2@18d613832
  corefx_510ab9e6433ff0df29e0ae080140f4f2c1b69abd/revision-cbdb97070c2@39a45ceef
  corefx_510ab9e6433ff0df29e0ae080140f4f2c1b69abd/revision-cbdb97070c2@18d613832
  EntityFrameworkCore_a2f17c7f90b63f57b56fc2a336e115b7b1e960d3/revision-cbdb97070c2@39a45ceef
  EntityFrameworkCore_a2f17c7f90b63f57b56fc2a336e115b7b1e960d3/revision-cbdb97070c2@18d613832
  runuo/revision-cbdb97070c2@39a45ceef
  runuo/revision-cbdb97070c2@18d613832
  coreclr/revision-cbdb97070c2@39a45ceef
  coreclr/revision-cbdb97070c2@18d613832
  PowerShell_ba94c6cfeb93d66bd30748ffa181d2e6d541351e/revision-cbdb97070c2@39a45ceef
  PowerShell_ba94c6cfeb93d66bd30748ffa181d2e6d541351e/revision-cbdb97070c2@18d613832
  pythonnet_51a186858096577aef6a860656523d3194eee9a8/revision-cbdb97070c2@39a45ceef
  pythonnet_51a186858096577aef6a860656523d3194eee9a8/revision-cbdb97070c2@18d613832
  roslyn_6a52142bef1db0a0936c483d13f426acae335555/revision-cbdb97070c2@39a45ceef
  roslyn_6a52142bef1db0a0936c483d13f426acae335555/revision-cbdb97070c2@18d613832
  nhibernate-core_7997e22b1d999fecacd91dc8664f5132001cebfb/revision-cbdb97070c2@39a45ceef
  nhibernate-core_7997e22b1d999fecacd91dc8664f5132001cebfb/revision-cbdb97070c2@18d613832
```
# Result summary

Legend:
- `+`: Added results
- `-`: Removed results
- `(+)`: Added results, in classified sources (heuristic)
- `(-)`: Removed results, in classified sources (heuristic)

| | + | - | (+) | (-) |
| --- | :---: | :---: | :---: | :---: |
| TOTAL |  |  |  |  |
# + results (0) (UNCLASSIFIED)
# - results (0) (UNCLASSIFIED)
# + results (0) (CLASSIFIED)
# - results (0) (CLASSIFIED)
